### PR TITLE
Action that runs quarto render to check for broken links and errors

### PR DESCRIPTION
**Pull request by noosheenv**
## Internal Notes for Reviewers
I made a Github action to run `quarto render` to check for errors and broken links every time you push a commit.

The file is called `render-site.yaml` and resides in the `.github/workflows` folder.

The first few steps installs any necessary dependencies, such as Quarto and its associated extensions by using `quarto install extension`. It also tells the action to run when someone either makes a new PR or pushes something.

Then, it runs `quarto render` and checks for any immediate errors, like if you had something really off in the `_quarto.yml` file. Here's what it looks like when you try to push something with a big error:
image
It shows the output and also displays a message saying Quarto failed immediately.

If there are no big errors, `quarto render` will run in the background, which takes approximately 5 minutes. The workflow will then check the output for any instances of warnings or errors. If there are warnings, usually for broken links, the check will fail and display where the warning is. Here is an example:
image
You can find a detailed view like the screenshot above by clicking on **Details** next to the Quarto render check like this:
image

If there are no errors or warnings, the check will pass.

**Other changes:**
There were still some issues with broken links relating to the name of the `NLP_and_LLM` folder, so some commits fix that. Also, some release notes were referencing our hacky `NLP_and_LLM` folder, so upon fixing the folder name, the links were updated to reference `nlp_and_llm`.

### CODE REVIEW

File: .github/workflows/render-site.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          quarto render &> render_errors.log || { 
+          echo "Quarto render failed immediately";
+          cat render_errors.log;
+          exit 1;
+          }
+
+    - name: Check for render errors 
+      run: |
+        if grep -q 'WARN:\|ERROR:' site/render_errors.log; then
+          echo "Warnings or errors detected during render"
```
Comment (nrichers) [inline .github/workflows/render-site.yaml:44]:
```suggestion
          echo "Warnings or errors detected during Quarto render"
```

File: .github/workflows/render-site.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          cat site/render_errors.log
+          exit 1
+        else
+          echo "No warnings or errors detected during render"
```
Comment (nrichers) [inline .github/workflows/render-site.yaml:48]:
```suggestion
          echo "No warnings or errors detected during Quarto render"
```

File: .github/workflows/render-site.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
+name: Check rendered site for broken links
+
+on:
+  push:
+    branches:
+      - main
```
Comment (nrichers) [inline .github/workflows/render-site.yaml:6]: Does your workflow run fine without these lines? We don't allow pushing directly against `main` so this should technically not do anything. If so, I'd delete.

File: .github/workflows/render-site.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  pull_request:
+    branches:
```
Comment (nrichers) [inline .github/workflows/render-site.yaml:8]: I'd add a statement here that exludes draft PRs (runs the workflow only for PRs that are ready for review). You can look at https://github.com/validmind/documentation/blob/main/.github/workflows/release_notes_check.yml#L5 for an example of how that's done, the `ready_for_review` bit specifically.

File: .github/workflows/render-site.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+
+    - name: Set up Quarto
```
Comment (nrichers) [inline .github/workflows/render-site.yaml:19]: I'd try something like this to always grab the latest but you will need to test this actually works: 

```
- name: Set up Quarto
  run: |
    latest_version=$(curl -s https://api.github.com/repos/quarto-dev/quarto-cli/releases/latest | grep 'tag_name' | cut -d\" -f4)
    wget https://github.com/quarto-dev/quarto-cli/releases/download/${latest_version}/quarto-${latest_version}-linux-amd64.deb
    sudo dpkg -i quarto-${latest_version}-linux-amd64.deb
```

File: .github/workflows/render-site.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        wget https://github.com/quarto-dev/quarto-cli/releases/download/v1.5.53/quarto-1.5.53-linux-amd64.deb
+        sudo dpkg -i quarto-1.5.53-linux-amd64.deb
+# idk about this
+
+    - name: Install dependencies
+      run: |
+        if [ -f _extensions/quarto-ext.yml ]; then
+          quarto install extension
+        fi
+# idk about this one too
```
Comment (nrichers) [inline .github/workflows/render-site.yaml:30]: I think this is OK but I'm not convinced it does anything, as we don't seem to have that file in our folder structure. The extensions we need are all already installed and would be included when the repo gets cloned as part of the CI check.

**Review by nrichers — COMMENTED** [review]
Nice work, @noosheenv! Kudos on your first infrastructure PR. This looks great and also catches other errors, e.g. when an include statement is broken. 🚀 

The comments I left are mainly to tweak a few things here and there. Let me know when you've made those changes and I will retest.

File: .github/workflows/render-site.yaml (hunk 5 above)
Comment (noosheenv) [reply .github/workflows/render-site.yaml]: This one was giving broken links, so I made it check the API site for the whole URL:
```
- name: Get latest Quarto release URL
      id: get-quarto-url
      run: |
        API_URL="https://api.github.com/repos/quarto-dev/quarto-cli/releases/latest"
        DOWNLOAD_URL=$(curl -s $API_URL | jq -r '.assets[] | select(.name | endswith("linux-amd64.deb")).browser_download_url')
        echo "DOWNLOAD_URL=$DOWNLOAD_URL" >> $GITHUB_ENV

    - name: Download and install Quarto
      run: |
        wget ${{ env.DOWNLOAD_URL }} -O quarto-latest-linux-amd64.deb
        sudo dpkg -i quarto-latest-linux-amd64.deb
```

**Comment by noosheenv** [comment]
Summary of changes:
- Got rid of the on push thing, and just had the same line as the release notes check: 
```
on:
  pull_request:
    types: [opened, synchronize, labeled, unlabeled, edited, ready_for_review]
```
- Learned that Github runs workflows only if PR is able to merge first, and that's why the tests weren't running, not because of this `on: pull_request` thing
- Got Chatgpt to help me retrieve a working download link of the latest Quarto version, building off what you had (almost there but was giving error 404)
- Cleaned up comments and fixed suggested wording changes

File: .github/workflows/render-site.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
+name: Run quarto render
+
+on:
+  pull_request:
+    types: [opened, synchronize, labeled, unlabeled, edited, ready_for_review]
```
Comment (nrichers) [inline .github/workflows/render-site.yaml:5]:
```suggestion
    types: [opened, synchronize, edited, ready_for_review]
```

You don't need the label function here. In fact, given the new CI check takes several minutes to run, we should minimize how often we trigger.

**Review by nrichers — COMMENTED** [review]
@noosheenv re-reviewed and this looks good, except the new CI check runs even when the PR is in draft mode: 

image

You can see in the screenshot that the release notes check was skipped — that's how it should be.

**Comment by noosheenv** [comment]
@nrichers I seeeeee I needed an if statement under the build to skip if the PR is a draft. Should be resolved with this commit: cbabceb

**Review by nrichers — APPROVED** [review]
LGTM! (Looks GREAT to me.) Retested and your new GitHub workflow works as expected, including skipping when the PR is in draft mode. SHIP IT. 🚢

**Pull request merged**